### PR TITLE
Fixed tests and added example of fabric with HPU

### PR DESCRIPTION
**Pull request by jyothisambolu**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: examples/fabric/image_classifier/train_fabric.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        batch_size=hparams.batch_size,
+    )
+    test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=hparams.batch_size)
+
+    # don't forget to call `setup_dataloaders` to prepare for dataloaders for distributed training.
+    train_loader, test_loader = fabric.setup_dataloaders(train_loader, test_loader)
+
+    model = Net()  # remove call to .to(device)
```
Comment (jerome-habana) [inline examples/fabric/image_classifier/train_fabric.py:102]: remove commented code

File: examples/fabric/image_classifier/train_fabric.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    optimizer = optim.Adadelta(model.parameters(), lr=hparams.lr)
+
+    # don't forget to call `setup` to prepare for model / optimizer for distributed training.
```
Comment (jerome-habana) [inline examples/fabric/image_classifier/train_fabric.py:105]: why do we need this helper comments ?

File: examples/fabric/image_classifier/train_fabric.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    # use torchmetrics instead of manually computing the accuracy
+    test_acc = Accuracy(task="multiclass", num_classes=10).to(fabric.device)
+
+    # EPOCH LOOP
+    for epoch in range(1, hparams.epochs + 1):
+        # TRAINING LOOP
+        model.train()
+        for batch_idx, (data, target) in enumerate(train_loader):
+            # NOTE: no need to call `.to(device)` on the data, target
```
Comment (jerome-habana) [inline examples/fabric/image_classifier/train_fabric.py:119]: ditto

File: examples/fabric/image_classifier/train_fabric.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        model.eval()
+        test_loss = 0
+        with torch.no_grad():
+            for data, target in test_loader:
+                # NOTE: no need to call `.to(device)` on the data, target
+                output = model(data)
+                test_loss += F.nll_loss(output, target, reduction="sum").item()
+
+                # WITHOUT TorchMetrics
+                # pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
```
Comment (jerome-habana) [inline examples/fabric/image_classifier/train_fabric.py:151]: ??

File: examples/fabric/image_classifier/train_fabric.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    # Example:
+    # lightning run model image_classifier.py accelerator=cuda --epochs=3
+    parser = argparse.ArgumentParser(description="Fabric MNIST Example")
+    parser.add_argument(
+        "--batch-size", type=int, default=64, metavar="N", help="input batch size for training (default: 64)"
+    )
+    parser.add_argument("--epochs", type=int, default=14, metavar="N", help="number of epochs to train (default: 14)")
+    parser.add_argument("--lr", type=float, default=1.0, metavar="LR", help="learning rate (default: 1.0)")
+    parser.add_argument("--gamma", type=float, default=0.7, metavar="M", help="Learning rate step gamma (default: 0.7)")
+    parser.add_argument("--dry-run", action="store_true", default=False, help="quickly check a single pass")
```
Comment (jerome-habana) [inline examples/fabric/image_classifier/train_fabric.py:186]: why dry-run in example ?

File: src/lightning_habana/fabric/strategies/parallel.py [hunk 6]
```diff
@@ -120,6 +120,20 @@ def backward(self, tensor: Tensor, module: Optional[Module], *args: Any, **kwarg
             # Break lazy accumulation of graph after fwd+bwd
             htcore.mark_step()
 
+    def setup_module(self, module: Module) -> Module:
+        """Performs setup for the model, e.g., by wrapping it by another class."""
+        # Fabric doesn't support nn.Module with wrapped attributes currently.
```
Comment (jerome-habana) [inline src/lightning_habana/fabric/strategies/parallel.py:125]: We need to point to the GH issue here

File: examples/fabric/image_classifier/train_fabric.py (hunk 1 above)
Comment (jyothisambolu) [reply examples/fabric/image_classifier/train_fabric.py]: Fixed

File: examples/fabric/image_classifier/train_fabric.py (hunk 2 above)
Comment (jyothisambolu) [reply examples/fabric/image_classifier/train_fabric.py]: Removed helper comments

File: examples/fabric/image_classifier/train_fabric.py (hunk 5 above)
Comment (jyothisambolu) [reply examples/fabric/image_classifier/train_fabric.py:156]: dry-run is used to quickly check single pass of training loop. It is used in the lightning example as well.

File: src/lightning_habana/fabric/strategies/parallel.py (hunk 6 above)
Comment (jyothisambolu) [reply src/lightning_habana/fabric/strategies/parallel.py:125]: Added

**Review by Borda — APPROVED** [review]

**Review by jerome-habana — APPROVED** [review]

**Pull request merged**